File: Practice Problems/C++/findPairsWithDifference.cpp
#include <stdio.h>
#include <stdlib.h>
#include <unordered_set>
#include <vector>
#include <iostream>
using namespace std;

struct Pair{
	int a;
	int b;

};


//given an array arr of size n, return a  vector of pairs with a difference
//of difference. Ex [1,2,4,5,8] with a difference of 3 would return a vector
//[(1,4), (2,5), (5,8)]
vector<Pair> findPairsWithDifference(int n, int difference, int arr[]){
	unordered_set<int> setArr;
	vector<Pair> result;

	if(difference == 0){
		Pair temp;
		for(int i = 0; i < n; i++){
			if(setArr.find(arr[i]) != setArr.end()){
				temp.a = arr[i];
				temp.b = arr[i];
				result.push_back(temp);
			} else {
				setArr.insert(arr[i]);
			}
		}
		return result;
	}


	for(int i=0; i < n; i++)
		setArr.insert(arr[i]);

	unordered_set<int>::iterator it = setArr.begin();
	Pair temp;
	while(it != setArr.end()){
		unordered_set<int>::iterator node = setArr.find((*it) + difference);
		if(node != setArr.end()){
				temp.a = *it;
				temp.b = *node;
			result.push_back(temp);
		}
		it++;
	}
}
void printVectorPair(vector<Pair> vec){
	for(int i=0; i < vec.size(); i++){
		if(i == vec.size()-1){
			cout << "( " << vec[i].a <<", " << vec[i].b << ")\n";
		} else {
	 		cout << "( " << vec[i].a <<", " << vec[i].b << "), ";
	 	}
	}
}	

int main(int argc, char* argv[]){




	int test1[5] = {1,2,4,5,8};
	int test2[7] = {1,1,1,4,8,9,9};
	printVectorPair(findPairsWithDifference(5, 3, test1));
	printVectorPair(findPairsWithDifference(7, 0, test2));
	printVectorPair(findPairsWithDifference(5, -3, test1));
}

File: Practice Problems/Python/findPairsWithDifference.py
from sets import Set

class Pair:
	a = 1
	b = 2

	def __init__(self, myA, myB):
		self.a = myA
		self.b = myB

	def __repr__(self):
		return "(" + str(self.a) + ", " + str(self.b) + ")"
	def __str__(self):
		return "A: " + self.a + " B: " + self.b + " "


def findPairsWithDifference(difference, arr):
	setArr = Set([])
	result = []
	if difference == 0:
		for i in arr:
			if i in setArr:
				result.append(Pair(i,i))
			else:
				setArr.add(i)
		return result
	for i in arr:
		setArr.add(i)

	for i in setArr:
		if (i + difference) in arr:
			result.append(Pair(i, i + difference))
	return result

print findPairsWithDifference(3, [1,2,3,4,5,6])
print findPairsWithDifference(0, [1,1,2,3,4,5,6,9,9,9])
print findPairsWithDifference(-1, [1,1,2,3,4,5,6,9,9,9])